File: web/src/api/redeem.ts
import { authFetch } from './client'
import type { UserSubscriptionResponse } from './subscription'

const API_BASE = '/api'

async function fetchJson<T>(url: string, options?: RequestInit): Promise<T> {
  const res = await authFetch(url, options)
  let data: unknown

  try {
    data = await res.json()
  } catch {
    data = undefined
  }

  if (!res.ok) {
    const payload = data as { error?: string; message?: string } | undefined
    throw new Error(payload?.error || payload?.message || `请求失败 (${res.status})`)
  }

  return data as T
}

export type RedeemCodeMode = 'single_use' | 'shared'
export type RedeemCodeStatus = 'active' | 'disabled' | 'consumed'
export type RedeemRedemptionStatus = 'success' | 'rejected'

export interface RedeemCampaign {
  id: string
  name: string
  description: string
  codeMode: RedeemCodeMode
  sharedCode?: string
  sharedCodeMask?: string
  subscriptionPlanId: string
  subscriptionPlanName: string
  subscriptionDurationDays: number
  balanceMicros: number
  totalRedemptionsLimit: number
  redeemedCount: number
  perUserLimit: number
  startsAt: string | null
  endsAt: string | null
  enabled: boolean
  codeCount: number
  batchCount: number
  createdAt: string
  updatedAt: string
}

export interface RedeemCampaignRequest {
  name: string
  description: string
  codeMode: RedeemCodeMode
  sharedCode?: string
  subscriptionPlanId: string
  subscriptionDurationDays: number
  balanceMicros: number
  totalRedemptionsLimit: number
  perUserLimit: number
  startsAt?: string
  endsAt?: string
  enabled: boolean
}

export interface RedeemBatch {
  id: string
  campaignId: string
  campaignName: string
  name: string
  prefix: string
  codeCount: number
  codeLength: number
  createdAt: string
  updatedAt: string
}

export interface RedeemBatchRequest {
  name: string
  prefix: string
  codeCount: number
  codeLength: number
}

export interface RedeemCode {
  id: string
  campaignId: string
  campaignName: string
  batchId?: string | null
  batchName?: string
  codeMode: RedeemCodeMode
  codeValue: string
  codeMask: string
  status: RedeemCodeStatus
  maxRedemptions: number
  redeemedCount: number
  lastRedeemedAt: string | null
  createdAt: string
  updatedAt: string
}

export interface RedeemRedemption {
  id: string
  campaignId?: string | null
  campaignName: string
  codeId?: string | null
  userId: string
  username: string
  codeMask: string
  subscriptionPlanId: string
  subscriptionPlanName: string
  subscriptionDurationDays: number
  balanceMicros: number
  status: RedeemRedemptionStatus
  failureReason: string
  grantedSubscriptionId: string
  grantedExpiresAt: string | null
  balanceAfterMicros: number
  createdAt: string
}

export interface RedeemResult {
  message: string
  campaign: RedeemCampaign | null
  redemption: RedeemRedemption | null
  currentSubscription?: UserSubscriptionResponse | null
  balanceMicros: number
  balanceUsd: string
}

export async function redeemCode(code: string): Promise<RedeemResult> {
  return fetchJson<RedeemResult>(`${API_BASE}/me/redeem`, {
    method: 'POST',
    body: JSON.stringify({ code }),
  })
}

export async function listMyRedeemRecords(): Promise<RedeemRedemption[]> {
  const data = await fetchJson<{ items: RedeemRedemption[] }>(`${API_BASE}/me/redeem/records`)
  return data.items || []
}

export async function listRedeemCampaigns(): Promise<RedeemCampaign[]> {
  const data = await fetchJson<{ items: RedeemCampaign[] }>(`${API_BASE}/admin/redeem/campaigns`)
  return data.items || []
}

export async function createRedeemCampaign(payload: RedeemCampaignRequest): Promise<RedeemCampaign> {
  return fetchJson<RedeemCampaign>(`${API_BASE}/admin/redeem/campaigns`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function updateRedeemCampaign(id: string, payload: RedeemCampaignRequest): Promise<RedeemCampaign> {
  return fetchJson<RedeemCampaign>(`${API_BASE}/admin/redeem/campaigns/${encodeURIComponent(id)}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export async function deleteRedeemCampaign(id: string): Promise<void> {
  await fetchJson<{ message: string }>(`${API_BASE}/admin/redeem/campaigns/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  })
}

export async function listRedeemBatches(campaignId = ''): Promise<RedeemBatch[]> {
  const query = campaignId ? `?campaignId=${encodeURIComponent(campaignId)}` : ''
  const data = await fetchJson<{ items: RedeemBatch[] }>(`${API_BASE}/admin/redeem/batches${query}`)
  return data.items || []
}

export async function createRedeemBatch(campaignId: string, payload: RedeemBatchRequest): Promise<{ batch: RedeemBatch; codes: string[] }> {
  return fetchJson<{ batch: RedeemBatch; codes: string[] }>(`${API_BASE}/admin/redeem/campaigns/${encodeURIComponent(campaignId)}/batches`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function downloadRedeemBatchCSV(batchId: string): Promise<Blob> {
  const res = await authFetch(`${API_BASE}/admin/redeem/batches/${encodeURIComponent(batchId)}/export`)
  if (!res.ok) {
    let message = '导出失败'
    try {
      const payload = await res.json()
      message = payload.error || message
    } catch {
      // ignore
    }
    throw new Error(message)
  }
  return res.blob()
}

export interface RedeemCodeFilters {
  campaignId?: string
  batchId?: string
  status?: RedeemCodeStatus | ''
  keyword?: string
  limit?: number
}

export async function listRedeemCodes(filters: RedeemCodeFilters = {}): Promise<RedeemCode[]> {
  const params = new URLSearchParams()
  if (filters.campaignId) params.set('campaignId', filters.campaignId)
  if (filters.batchId) params.set('batchId', filters.batchId)
  if (filters.status) params.set('status', filters.status)
  if (filters.keyword) params.set('keyword', filters.keyword)
  if (filters.limit) params.set('limit', String(filters.limit))
  const query = params.toString()
  const data = await fetchJson<{ items: RedeemCode[] }>(`${API_BASE}/admin/redeem/codes${query ? `?${query}` : ''}`)
  return data.items || []
}

export async function setRedeemCodeEnabled(id: string, enabled: boolean): Promise<void> {
  await fetchJson<{ message: string }>(`${API_BASE}/admin/redeem/codes/${encodeURIComponent(id)}/enabled`, {
    method: 'PATCH',
    body: JSON.stringify({ enabled }),
  })
}

export interface RedeemRedemptionFilters {
  campaignId?: string
  status?: RedeemRedemptionStatus | ''
  username?: string
  limit?: number
}

export async function listRedeemRedemptions(filters: RedeemRedemptionFilters = {}): Promise<RedeemRedemption[]> {
  const params = new URLSearchParams()
  if (filters.campaignId) params.set('campaignId', filters.campaignId)
  if (filters.status) params.set('status', filters.status)
  if (filters.username) params.set('username', filters.username)
  if (filters.limit) params.set('limit', String(filters.limit))
  const query = params.toString()
  const data = await fetchJson<{ items: RedeemRedemption[] }>(`${API_BASE}/admin/redeem/redemptions${query ? `?${query}` : ''}`)
  return data.items || []
}

File: internal/model/redeem.go
package model

import "time"

type RedeemCodeMode string

const (
	RedeemCodeModeSingleUse RedeemCodeMode = "single_use"
	RedeemCodeModeShared    RedeemCodeMode = "shared"
)

type RedeemCodeStatus string

const (
	RedeemCodeStatusActive   RedeemCodeStatus = "active"
	RedeemCodeStatusDisabled RedeemCodeStatus = "disabled"
	RedeemCodeStatusConsumed RedeemCodeStatus = "consumed"
)

type RedeemRedemptionStatus string

const (
	RedeemRedemptionStatusSuccess  RedeemRedemptionStatus = "success"
	RedeemRedemptionStatusRejected RedeemRedemptionStatus = "rejected"
)

type RedeemCodeSourceType string

const (
	RedeemCodeSourceTypeCampaign       RedeemCodeSourceType = "campaign"
	RedeemCodeSourceTypeFree           RedeemCodeSourceType = "free"
	RedeemCodeSourceTypePurchaseOrder  RedeemCodeSourceType = "purchase_order"
	RedeemCodeSourceTypeLegacyDelivery RedeemCodeSourceType = "legacy_delivery"
)

type RedeemCampaign struct {
	ID                       string         `json:"id"`
	Name                     string         `json:"name"`
	Description              string         `json:"description"`
	CodeMode                 RedeemCodeMode `json:"codeMode"`
	SubscriptionPlanID       string         `json:"subscriptionPlanId"`
	SubscriptionDurationDays int            `json:"subscriptionDurationDays"`
	BalanceMicros            int64          `json:"balanceMicros"`
	TotalRedemptionsLimit    int            `json:"totalRedemptionsLimit"`
	RedeemedCount            int            `json:"redeemedCount"`
	PerUserLimit             int            `json:"perUserLimit"`
	StartsAt                 *time.Time     `json:"startsAt"`
	EndsAt                   *time.Time     `json:"endsAt"`
	Enabled                  bool           `json:"enabled"`
	CreatedAt                time.Time      `json:"createdAt"`
	UpdatedAt                time.Time      `json:"updatedAt"`
}

type RedeemCampaignRequest struct {
	Name                     string         `json:"name" binding:"required,min=1,max=64"`
	Description              string         `json:"description" binding:"max=160"`
	CodeMode                 RedeemCodeMode `json:"codeMode" binding:"required,oneof=single_use shared"`
	SharedCode               string         `json:"sharedCode" binding:"max=64"`
	SubscriptionPlanID       string         `json:"subscriptionPlanId"`
	SubscriptionDurationDays int            `json:"subscriptionDurationDays" binding:"min=0,max=3650"`
	BalanceMicros            int64          `json:"balanceMicros" binding:"min=0"`
	TotalRedemptionsLimit    int            `json:"totalRedemptionsLimit" binding:"min=0,max=1000000"`
	PerUserLimit             int            `json:"perUserLimit" binding:"min=1,max=10000"`
	StartsAt                 *time.Time     `json:"startsAt"`
	EndsAt                   *time.Time     `json:"endsAt"`
	Enabled                  bool           `json:"enabled"`
}

type RedeemCampaignResponse struct {
	ID                       string         `json:"id"`
	Name                     string         `json:"name"`
	Description              string         `json:"description"`
	CodeMode                 RedeemCodeMode `json:"codeMode"`
	SharedCode               string         `json:"sharedCode,omitempty"`
	SharedCodeMask           string         `json:"sharedCodeMask,omitempty"`
	SubscriptionPlanID       string         `json:"subscriptionPlanId"`
	SubscriptionPlanName     string         `json:"subscriptionPlanName"`
	SubscriptionDurationDays int            `json:"subscriptionDurationDays"`
	BalanceMicros            int64          `json:"balanceMicros"`
	TotalRedemptionsLimit    int            `json:"totalRedemptionsLimit"`
	RedeemedCount            int            `json:"redeemedCount"`
	PerUserLimit             int            `json:"perUserLimit"`
	StartsAt                 *time.Time     `json:"startsAt"`
	EndsAt                   *time.Time     `json:"endsAt"`
	Enabled                  bool           `json:"enabled"`
	CodeCount                int            `json:"codeCount"`
	BatchCount               int            `json:"batchCount"`
	CreatedAt                time.Time      `json:"createdAt"`
	UpdatedAt                time.Time      `json:"updatedAt"`
}

type RedeemCodeBatch struct {
	ID         string    `json:"id"`
	CampaignID string    `json:"campaignId"`
	Name       string    `json:"name"`
	Prefix     string    `json:"prefix"`
	CodeCount  int       `json:"codeCount"`
	CodeLength int       `json:"codeLength"`
	CreatedAt  time.Time `json:"createdAt"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

type RedeemCodeBatchRequest struct {
	Name       string `json:"name" binding:"required,min=1,max=64"`
	Prefix     string `json:"prefix" binding:"max=16"`
	CodeCount  int    `json:"codeCount" binding:"required,min=1,max=5000"`
	CodeLength int    `json:"codeLength" binding:"required,min=6,max=32"`
}

type RedeemCodeBatchResponse struct {
	ID           string    `json:"id"`
	CampaignID   string    `json:"campaignId"`
	CampaignName string    `json:"campaignName"`
	Name         string    `json:"name"`
	Prefix       string    `json:"prefix"`
	CodeCount    int       `json:"codeCount"`
	CodeLength   int       `json:"codeLength"`
	CreatedAt    time.Time `json:"createdAt"`
	UpdatedAt    time.Time `json:"updatedAt"`
}

type RedeemCode struct {
	ID                       string               `json:"id"`
	CampaignID               *string              `json:"campaignId"`
	BatchID                  *string              `json:"batchId"`
	SourceType               RedeemCodeSourceType `json:"sourceType"`
	SourceRefID              string               `json:"sourceRefId"`
	CodeValue                string               `json:"codeValue"`
	CodeHash                 string               `json:"-"`
	CodeMask                 string               `json:"codeMask"`
	SubscriptionPlanID       *string              `json:"subscriptionPlanId"`
	SubscriptionDurationDays int                  `json:"subscriptionDurationDays"`
	BalanceMicros            int64                `json:"balanceMicros"`
	RewardSnapshotJSON       string               `json:"rewardSnapshotJson,omitempty"`
	LegacySource             string               `json:"legacySource,omitempty"`
	LegacyRefID              string               `json:"legacyRefId,omitempty"`
	PerUserLimit             int                  `json:"perUserLimit"`
	StartsAt                 *time.Time           `json:"startsAt"`
	EndsAt                   *time.Time           `json:"endsAt"`
	Status                   RedeemCodeStatus     `json:"status"`
	MaxRedemptions           int                  `json:"maxRedemptions"`
	RedeemedCount            int                  `json:"redeemedCount"`
	LastRedeemedAt           *time.Time           `json:"lastRedeemedAt"`
	CreatedAt                time.Time            `json:"createdAt"`
	UpdatedAt                time.Time            `json:"updatedAt"`
}

type RedeemCodeResponse struct {
	ID                       string               `json:"id"`
	CampaignID               *string              `json:"campaignId"`
	CampaignName             string               `json:"campaignName"`
	BatchID                  *string              `json:"batchId"`
	BatchName                string               `json:"batchName,omitempty"`
	SourceType               RedeemCodeSourceType `json:"sourceType"`
	SourceRefID              string               `json:"sourceRefId"`
	CodeMode                 RedeemCodeMode       `json:"codeMode"`
	CodeValue                string               `json:"codeValue"`
	CodeMask                 string               `json:"codeMask"`
	SubscriptionPlanID       string               `json:"subscriptionPlanId"`
	SubscriptionPlanName     string               `json:"subscriptionPlanName"`
	SubscriptionDurationDays int                  `json:"subscriptionDurationDays"`
	BalanceMicros            int64                `json:"balanceMicros"`
	RewardSnapshotJSON       string               `json:"rewardSnapshotJson,omitempty"`
	LegacySource             string               `json:"legacySource,omitempty"`
	LegacyRefID              string               `json:"legacyRefId,omitempty"`
	PerUserLimit             int                  `json:"perUserLimit"`
	StartsAt                 *time.Time           `json:"startsAt"`
	EndsAt                   *time.Time           `json:"endsAt"`
	Status                   RedeemCodeStatus     `json:"status"`
	MaxRedemptions           int                  `json:"maxRedemptions"`
	RedeemedCount            int                  `json:"redeemedCount"`
	LastRedeemedAt           *time.Time           `json:"lastRedeemedAt"`
	CreatedAt                time.Time            `json:"createdAt"`
	UpdatedAt                time.Time            `json:"updatedAt"`
}

type ManualRedeemCodeRequest struct {
	CampaignID               string     `json:"campaignId"`
	CodeValue                string     `json:"codeValue" binding:"required,min=3,max=128"`
	SubscriptionPlanID       string     `json:"subscriptionPlanId"`
	SubscriptionDurationDays int        `json:"subscriptionDurationDays" binding:"min=0,max=3650"`
	BalanceMicros            int64      `json:"balanceMicros" binding:"min=0"`
	PerUserLimit             int        `json:"perUserLimit" binding:"min=1,max=10000"`
	MaxRedemptions           int        `json:"maxRedemptions" binding:"min=1,max=1000000"`
	StartsAt                 *time.Time `json:"startsAt"`
	EndsAt                   *time.Time `json:"endsAt"`
	Enabled                  bool       `json:"enabled"`
}

type RedeemRedemption struct {
	ID                       string                 `json:"id"`
	CampaignID               *string                `json:"campaignId"`
	CodeID                   *string                `json:"codeId"`
	UserID                   string                 `json:"userId"`
	Username                 string                 `json:"username"`
	CodeInput                string                 `json:"codeInput"`
	CodeMask                 string                 `json:"codeMask"`
	SubscriptionPlanID       string                 `json:"subscriptionPlanId"`
	SubscriptionDurationDays int                    `json:"subscriptionDurationDays"`
	BalanceMicros            int64                  `json:"balanceMicros"`
	RewardSnapshotJSON       string                 `json:"rewardSnapshotJson,omitempty"`
	Status                   RedeemRedemptionStatus `json:"status"`
	FailureReason            string                 `json:"failureReason"`
	GrantedSubscriptionID    string                 `json:"grantedSubscriptionId"`
	GrantedExpiresAt         *time.Time             `json:"grantedExpiresAt"`
	BalanceAfterMicros       int64                  `json:"balanceAfterMicros"`
	CreatedAt                time.Time              `json:"createdAt"`
}

type RedeemRedemptionResponse struct {
	ID                       string                 `json:"id"`
	CampaignID               *string                `json:"campaignId"`
	CampaignName             string                 `json:"campaignName"`
	CodeID                   *string                `json:"codeId"`
	UserID                   string                 `json:"userId"`
	Username                 string                 `json:"username"`
	CodeMask                 string                 `json:"codeMask"`
	SubscriptionPlanID       string                 `json:"subscriptionPlanId"`
	SubscriptionPlanName     string                 `json:"subscriptionPlanName"`
	SubscriptionDurationDays int                    `json:"subscriptionDurationDays"`
	BalanceMicros            int64                  `json:"balanceMicros"`
	RewardSnapshotJSON       string                 `json:"rewardSnapshotJson,omitempty"`
	Status                   RedeemRedemptionStatus `json:"status"`
	FailureReason            string                 `json:"failureReason"`
	GrantedSubscriptionID    string                 `json:"grantedSubscriptionId"`
	GrantedExpiresAt         *time.Time             `json:"grantedExpiresAt"`
	BalanceAfterMicros       int64                  `json:"balanceAfterMicros"`
	CreatedAt                time.Time              `json:"createdAt"`
}

type RedeemRequest struct {
	Code string `json:"code" binding:"required,min=3,max=128"`
}

type RedeemResultResponse struct {
	Message             string                    `json:"message"`
	Campaign            *RedeemCampaignResponse   `json:"campaign"`
	Redemption          *RedeemRedemptionResponse `json:"redemption"`
	CurrentSubscription *UserSubscriptionResponse `json:"currentSubscription,omitempty"`
	BalanceMicros       int64                     `json:"balanceMicros"`
	BalanceUsd          string                    `json:"balanceUsd"`
}
